src/urllib3/response.py

```diff
@@ -701,14 +702,46 @@ def read(
         flush_decoder = False
         fp_closed = getattr(self._fp, "closed", False)
 
+        # Reading more than 2 GiB (`int` max value in C) via SSL on some
+        # CPython 3.8, 3.9, and 3.10 versions leads to an overflow error
+        # that has to be prevented if `amt` or `self.length_remaining`
+        # indicate that it may happen.
+        c_int_max = (2**31) - 1
+        if (
+            (amt and amt > c_int_max)
+            or (self.length_remaining and self.length_remaining > c_int_max)
+        ) and (
+            (3, 8, 0, "a", 4) <= sys.version_info < (3, 9, 7)
+            or (3, 10, 0, "b", 1) <= sys.version_info < (3, 10, 0, "rc", 1)
+        ):
+
+            def read(amt: Optional[int] = None) -> bytes:
```

Comment:
While nested functions are nice, they're slower than normal functions since the function object gets built for every invocation and can be confusing. (Generally speaking, pretending Python is a functional language never ends well. Tail recursion is the other good example.)

Can you please:
 * place this nested function at the same level as the existing `read()` function? Maybe call it `_read_aux` or `_read_chunks`
 * place the `if` *inside* this new function?

Reply:
Done in 890474a